docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,555 @@
+# `MeshTrafficRoute`
+
+- Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4743
+Other issues covered:
+- https://github.com/kumahq/kuma/issues/4691
+- https://github.com/kumahq/kuma/issues/4690
+- https://github.com/kumahq/kuma/issues/3456
+- https://github.com/kumahq/kuma/issues/5309
+
+## Context and Problem Statement
+
+Now that we have the [new policy matching resources](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md), it's time to create one for traffic routung.
+We currently have both `TrafficRoute` and `MeshGatewayRoute`.
+
+This MADR does not address Gateway API integration per sebut does consider Gateway
+API's `HTTPRoute`, both as inspiration as well as to make sure we don't block
+integration.
+
+## Decision Drivers
+
+- `MeshGatewayRoute` and `TrafficRoute` have a large overlap in functionality
+  and the use cases for routing from a gateway and between services are similar
+- Gateway API has different resources per protocol
+
+## Considered Options
+
+- Two resources: `MeshGatewayRoute` and `MeshTrafficRoute`
+- Unified resource for services and gateway: `MeshTrafficRoute`
+- Unified resources but one per protocol, e.g. `MeshHTTPRoute`, `MeshTCPRoute`
+
+## Decision Outcome
+
+The new resource routes and alters requests depending on where it's coming from and where it's going to. Routes end up in the client-side Envoy config so we choose a `spec.to` resource.
+
+```yaml
+spec:
+ targetRef: ... # from where requests are coming
+ to:
+  - targetRef: ... # to where requests are going
+ default: ...
+```
+
+Each protocol gets its own resource:
+
+- `MeshHTTPRoute`
+- `MeshTCPRoute`
+
+which can be extended with others like GRPC.
+
+The route that's used for a given request is the most specific one that Kuma can
+apply given the `kuma.io/protocol` tag.
+
+### Positive Consequences
+
+- Route configuration can be very precisely targeted
+- Users need know only one set of resources for routing
+- A final spec that has the best of `MeshGatewayRoute` and `TrafficRoute`
+
+### New `MeshHTTPRoute` spec
+
+Routing configuration consists of a list of rules where each rule _matches_ some
+requests or traffic, modifies it in some way with _filters_ and then sends it to a _backend_.
+
+#### Matching
+
+The new route matches requests based on:
+
+- headers
+- path
+- method
+- query parameters
+
+##### Headers
+
+Headers can be tested for:
+
+- presence
+- absence
+- exact value
+- prefix value
+- regex value
+
+##### Spec
+
+A rule contains a list of match objects, each of which specifies a match for one
+or more of the above types:
+
+```yaml
+  matches:
+  - methods:
+    - GET
+    - POST
+    path:
+      type: RegularExpression|Exact|Prefix
+      value: ...
+    headers:
+      - type: RegularExpression|Exact|Prefix|Absent
```

Comment:
Is `Absent` an equivalent of `Not(Exact)`? Have we thought of `AbsentRegularExpression|AbsentExact|AbsentPrefix`?

Reply:
got it, makes sense